### PR TITLE
Fix train_step, test_step and tests for CLIP

### LINKED ISSUE #18670: TFClipModel fails to train because of None loss
### System Info

transformers version: 4.21.1
Platform: MacOS BigSur 11.6.7
Python version: 3.8.13
Huggingface_hub version: 0.8.1
Tensorflow version (GPU?): 2.7.3 (False)
Flax version (CPU?/GPU?/TPU?): not installed (NA)
Jax version: not installed
JaxLib version: not installed
Using GPU in script?: no
Using distributed or parallel set-up in script?: no

@patil-suraj

### Who can help?

@patil-suraj

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

This is the script run to attempt to fit the model to the example data. It is verbatim from the 4.21.1 docs with the addition of `model.fit`. The same error arose when working with my own project. The loss is always `None` as is `y` `y_pred` . Somewhere in the logic of https://github.com/huggingface/transformers/blob/132402d752044301b37e54405832738b16f49df6/src/transformers/modeling_tf_utils.py#L1116.

```python
from PIL import Image
import requests
from transformers import CLIPProcessor, TFCLIPModel
import tensorflow as tf

model = TFCLIPModel.from_pretrained("openai/clip-vit-base-patch32")
processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

inputs = processor(
    text=["a photo of a cat", "a photo of a dog"], images=[image, image], return_tensors="tf", padding=True
)

outputs = model(**inputs)

model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.00001))
model.fit(dict(inputs))
```

Results with an error of `zero gradient` because the gradients are all 0s, which I expect is caused by `y` and `y_pred` both being empty dicts.

### Expected behavior

Model.fit() on inputs from preprocessor completes a training step without error.

### PR DESCRIPTION
**Pull request by Rocketknight1**
CLIP models were not being tested correctly with `fit()` because the test skipped models without a `hf_compute_loss` method. This skip was added to skip base models like `TFBERTModel` that do not have specific output heads and losses. However, it also skips models like CLIP that do not use `compute_loss` / `hf_compute_loss` methods.

The new test checks whether the model's return type dataclass has a `loss` key, which is a more reliable check. Enabling this reveals the bug in `fit()` for TFClip, so this PR also includes fixes to `train_step` and `test_step` for CLIP and models like it that require `return_loss=True` to be passed, but do not set it by default.

Draft for now because this will likely flush out other bugs or cause other problems!

Fixes #18670.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
The tests passed. I'm as surprised as everyone else.

**Comment by ydshieh** [comment]
Thanks @Rocketknight1  for quick fix. I am wondering if it makes sense to wrap the loss computation for TFCLIP into a `hf_compute_loss`? 

I also see for `TFSegformerForSemanticSegmentation`, it defines `hf_compute_loss` under its own model class.

If doable, maybe it's good to add `TFSemanticSegmentationLoss`, `TFcontrastiveLoss` class etc.

**Comment by Rocketknight1** [comment]
@ydshieh I don't think that's necessary - the new check we use means we don't need to rely on `hf_compute_loss` being present anymore!

**Comment by Rocketknight1** [comment]
Just realized that using `return` instead of `continue` in the tests was skipping a lot of tests, which might have been why it was green so easily. Rerunning everything!

**Review by gante — APPROVED** [review]
Looks good 👍 And I agree, relying on a loss output is safer!

Will re-review after fixing errors (I see the latest commit, which replaces the `return`, broke some tests :D)

**Comment by Rocketknight1** [comment]
Further updates: Now that we're no longer incorrectly skipping tests, this turned up quite a few bugs! The main source of issues is that some more recent models are returning scalar losses, but both our `test_loss_computation` test and Keras `fit()` expect that the loss has some shape, even if that shape is `(1,)`.

File: tests/test_modeling_tf_common.py [hunk 1]
```diff
@@ -125,6 +126,25 @@ def _config_zero_init(config):
...
+    return_type = get_type_hints(model.call)
+    if "return" not in return_type:
+        return False
+    return_type = return_type["return"]
+    if hasattr(return_type, "__args__"):  # Awkward check for union because UnionType only turns up in 3.10
+        for type_annotation in return_type.__args__:
+            if inspect.isclass(type_annotation) and issubclass(type_annotation, ModelOutput):
+                field_names = [field.name for field in fields(type_annotation)]
+                if "loss" in field_names:
+                    return True
```
Comment (amyeroberts) [inline tests/test_modeling_tf_common.py:139]: You could escape earlier 🏃 
```suggestion
                for field_name in fields(type_annotation):
                    if field_name == "loss":
                        return True
```

File: tests/test_modeling_tf_common.py [hunk 2]
```diff
@@ -1333,72 +1353,74 @@ def test_loss_computation(self):
...
-
-                self.assertTrue(loss.shape.as_list() == expected_loss_size or loss.shape.as_list() == [1])
+            if not getattr(model, "hf_compute_loss", None) and not _return_type_has_loss(model):
+                continue
+            # The number of elements in the loss should be the same as the number of elements in the label
+            prepared_for_class = self._prepare_for_class(inputs_dict.copy(), model_class, return_labels=True)
+            added_label_names = sorted(list(prepared_for_class.keys() - inputs_dict.keys()), reverse=True)
+            if not added_label_names:
+                continue  # This test is only for models with easily-separable labels
+            added_label = prepared_for_class[added_label_names[0]]
```
Comment (amyeroberts) [inline tests/test_modeling_tf_common.py:1363]: Can we assumed there's only one added label? Not sure if it matters, but if we can and it does, can we add an assert above?

**Review by amyeroberts — COMMENTED** [review]
Overall structure looks good to me. Leaving some comments as I was 👀 , but just nits.

File: src/transformers/models/mobilebert/modeling_tf_mobilebert.py [hunk 3]
```diff
@@ -88,6 +88,37 @@
...
 
 
+class TFMobileBertPreTrainingLoss:
+    """
+    Loss function suitable for BERT-like pretraining, that is, the task of pretraining a language model by combining
+    NSP + MLM. .. note:: Any label of -100 will be ignored (along with the corresponding logits) in the loss
+    computation.
+    """
+
+    # Copied from transformers.models.bert.modeling_tf_bert.TFBertPreTrainingLoss.hf_compute_loss
```
Comment (sgugger) [inline src/transformers/models/mobilebert/modeling_tf_mobilebert.py:98]: Why not put the copied from above? How is the rest different?

**Review by sgugger — APPROVED** [review]
Left a nit, but LGTM! Thanks for working on this!

File: tests/test_modeling_tf_common.py [hunk 4]
```diff
@@ -125,6 +126,25 @@ def _config_zero_init(config):
...
+            if inspect.isclass(type_annotation) and issubclass(type_annotation, ModelOutput):
+                field_names = [field.name for field in fields(type_annotation)]
+                if "loss" in field_names:
+                    return True
+        return False
+    elif isinstance(return_type, tuple):
+        return False
+    elif isinstance(return_type, ModelOutput):
+        class_fields = fields(return_type)
+        return "loss" in class_fields
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/test_modeling_tf_common.py:145]: Ah, good spot, thank you!

File: tests/test_modeling_tf_common.py (hunk 1 above)
Comment (Rocketknight1) [reply tests/test_modeling_tf_common.py:139]: I think I'll leave this one as is - I'd have to rewrite it with `field_name.name` and it might get a little ugly!

File: src/transformers/models/mobilebert/modeling_tf_mobilebert.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/models/mobilebert/modeling_tf_mobilebert.py:98]: Fixed!
